bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+                            from,
+                            request_id,
+                            pkg,
+                            distributed_shares,
+                            &rack_secret,
+                            &mut self.envelopes,
+                        )
+                    }
+                    // Only LoadRackSecret and LearnReceived track shares so we
+                    // cannot get another variant back.
+                    //
+                    // If we get a `None` back we either haven't received enough
+                    // shares or we have a late response to a prior request. A
+                    // late response is very common, as we terminate the request
+                    // once a threshold is received but we may still receive
+                    // shares because we asked more than a threshold of peers
+                    // for a share. Logging this as an unexpected response would
+                    // be noisy and misleading.
+                    _ => Ok(None),
+                }
+            }
+            State::Learned { pkg } => {
+                validate_share(&from, &share, &pkg.share_digests)?;
+                match self.request_manager.on_share(from, request_id, share) {
+                    Some(TrackableRequest::LoadRackSecret { acks, .. }) => {
+                        let secret = combine_shares(&pkg.share, acks)?;
+                        Ok(Some(ApiOutput::RackSecret { request_id, secret }))
+                    }
+                    Some(TrackableRequest::LearnReceived { .. }) => {
+                        panic!(
+                            "Invariant violation: Learned members must not 
+                            accept 'Learn' requests"
+                        )
+                    }
+                    // Only LoadRackSecret and LearnReceived track shares so we
+                    // cannot get another variant back.
+                    //
+                    // If we get a `None` back we either haven't received enough
+                    // shares or we have a late response to a prior request. A
+                    // late response is very common, as we terminate the request
+                    // once a threshold is received but we may still receive
+                    // shares because we asked more than a threshold of peers
+                    // for a share. Logging this as an unexpected response would
+                    // be noisy and misleading.
+                    _ => Ok(None),
+                }
+            }
+        }
+    }
+
+    // Select the next peer in a round-robin fashion
+    fn next_peer(&self, current: &Baseboard) -> Option<Baseboard> {
+        let mut iter = self.connected_peers.range(current..);
+        let matched = iter.next() == Some(current);
+        iter.next()
+            .filter(|_| matched)
+            .or_else(|| self.connected_peers.first())
+            .cloned()
+    }
+}
+
+fn decrypt_and_send_share_response(
```

Comment:
Vague question: should this be a member function of _something_?  It's taking an awful lot of arguments...

Reply:
So, unfortunately, I moved this and a few other methods on `Fsm`into plain functions because of borrow checker complaints. While the borrow checker knows that it can access multiple mutable individual fields in self independently within the same method, it is not aware that it can do this across methods. However, I still wanted to move this functionality outside of `on_share` to make it a bit easier to read.

Although there may be better code structure possible, In the interest of urgency I think we should just leave this as is.